recipes/openscenegraph/all/conanfile.py

```diff
@@ -0,0 +1,529 @@
+        cmake.definitions["OSG_USE_REF_PTR_IMPLICIT_OUTPUT_CONVERSION"] = self.options.enable_ref_ptr_implicit_output_conversion
+        cmake.definitions["OSG_USE_REF_PTR_SAFE_DEREFERENCE"] = self.options.enable_ref_ptr_safe_dereference
+        cmake.definitions["OSG_ENVVAR_SUPPORTED"] = self.options.enable_envvar_support
+
+        if not self.options.enable_windowing_system:
+            cmake.definitions["OSG_WINDOWING_SYSTEM"] = None
+
+        cmake.definitions["BUILD_OSG_DEPRECATED_SERIALIZERS"] = self.options.enable_deprecated_serializers
+
+        cmake.definitions["OSG_TEXT_USE_FONTCONFIG"] = self.options.use_fontconfig
+
+        # Disable option dependencies unless we have a package for them
+        cmake.definitions["OSG_WITH_FREETYPE"] = self.options.with_freetype
+        cmake.definitions["OSG_WITH_OPENEXR"] = self.options.get_safe("with_openexr", False)
+        cmake.definitions["OSG_WITH_INVENTOR"] = False
+        cmake.definitions["OSG_WITH_JASPER"] = self.options.with_jasper
+        cmake.definitions["OSG_WITH_OPENCASCADE"] = False
+        cmake.definitions["OSG_WITH_COLLADA"] = False  # self.options.with_collada
+        cmake.definitions["OSG_WITH_FBX"] = False
+        cmake.definitions["OSG_WITH_ZLIB"] = self.options.with_zlib
+        cmake.definitions["OSG_WITH_GDAL"] = self.options.with_gdal
+        cmake.definitions["OSG_WITH_GTA"] = self.options.with_gta
+        cmake.definitions["OSG_WITH_CURL"] = self.options.with_curl
+        cmake.definitions["OSG_WITH_LIBVNCSERVER"] = False
+        cmake.definitions["OSG_WITH_DCMTK"] = self.options.get_safe("with_dcmtk", False)
+        cmake.definitions["OSG_WITH_FFMPEG"] = False
+        cmake.definitions["OSG_WITH_GSTREAMER"] = False  # self.options.with_gstreamer
+        cmake.definitions["OSG_WITH_DIRECTSHOW"] = False
+        cmake.definitions["OSG_WITH_SDL"] = False
+        cmake.definitions["OSG_WITH_POPPLER"] = False
+        cmake.definitions["OSG_WITH_RSVG"] = False
+        cmake.definitions["OSG_WITH_NVTT"] = False
+        cmake.definitions["OSG_WITH_ASIO"] = False  # self.options.get_safe("with_asio", False)
+        cmake.definitions["OSG_WITH_ZEROCONF"] = False
+        cmake.definitions["OSG_WITH_LIBLAS"] = False
+        cmake.definitions["OSG_WITH_GIF"] = self.options.get_safe("with_gif", False)
+        cmake.definitions["OSG_WITH_JPEG"] = self.options.get_safe("with_jpeg", False)
+        cmake.definitions["OSG_WITH_PNG"] = self.options.get_safe("with_png", False)
+        cmake.definitions["OSG_WITH_TIFF"] = self.options.with_tiff
+
+        if self.settings.os == "Windows":
+            # osg has optional quicktime support on Windows
+            cmake.definitions["CMAKE_DISABLE_FIND_PACKAGE_QuickTime"] = True
+
+        cmake.definitions["OSG_MSVC_VERSIONED_DLL"] = False
+
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        CMake(self).install()
```

Comment:
```suggestion
    def _patch_sources(self):
        for patch in self.conan_data["patches"].get(self.version, []):
            tools.patch(**patch)

        for package in ("Fontconfig", "Freetype", "GDAL", "GIFLIB", "GTA", "Jasper", "OpenEXR"):
            # Prefer conan's find package scripts over osg's
            os.unlink(os.path.join(self._source_subfolder, "CMakeModules", "Find{}.cmake".format(package)))

    def _configure_cmake(self):
        if self._cmake:
            return self._cmake
        cmake = CMake(self)

        cmake.definitions["USE_3RDPARTY_BIN"] = False

        cmake.definitions["DYNAMIC_OPENSCENEGRAPH"] = self.options.shared
        cmake.definitions["DYNAMIC_OPENTHREADS"] = self.options.shared

        cmake.definitions["BUILD_OSG_APPLICATIONS"] = self.options.build_applications
        cmake.definitions["BUILD_OSG_EXAMPLES"] = self.options.build_examples

        cmake.definitions["OSG_NOTIFY_DISABLED"] = not self.options.enable_notify
        cmake.definitions["OSG_USE_DEPRECATED_API"] = self.options.enable_deprecated_api
        cmake.definitions["OSG_PROVIDE_READFILE"] = self.options.enable_readfile
        cmake.definitions["OSG_USE_REF_PTR_IMPLICIT_OUTPUT_CONVERSION"] = self.options.enable_ref_ptr_implicit_output_conversion
        cmake.definitions["OSG_USE_REF_PTR_SAFE_DEREFERENCE"] = self.options.enable_ref_ptr_safe_dereference
        cmake.definitions["OSG_ENVVAR_SUPPORTED"] = self.options.enable_envvar_support

        if not self.options.enable_windowing_system:
            cmake.definitions["OSG_WINDOWING_SYSTEM"] = None

        cmake.definitions["BUILD_OSG_DEPRECATED_SERIALIZERS"] = self.options.enable_deprecated_serializers

        cmake.definitions["OSG_TEXT_USE_FONTCONFIG"] = self.options.use_fontconfig

        # Disable option dependencies unless we have a package for them
        cmake.definitions["OSG_WITH_FREETYPE"] = self.options.with_freetype
        cmake.definitions["OSG_WITH_OPENEXR"] = self.options.get_safe("with_openexr", False)
        cmake.definitions["OSG_WITH_INVENTOR"] = False
        cmake.definitions["OSG_WITH_JASPER"] = self.options.with_jasper
        cmake.definitions["OSG_WITH_OPENCASCADE"] = False
        cmake.definitions["OSG_WITH_COLLADA"] = False  # self.options.with_collada
        cmake.definitions["OSG_WITH_FBX"] = False
        cmake.definitions["OSG_WITH_ZLIB"] = self.options.with_zlib
        cmake.definitions["OSG_WITH_GDAL"] = self.options.with_gdal
        cmake.definitions["OSG_WITH_GTA"] = self.options.with_gta
        cmake.definitions["OSG_WITH_CURL"] = self.options.with_curl
        cmake.definitions["OSG_WITH_LIBVNCSERVER"] = False
        cmake.definitions["OSG_WITH_DCMTK"] = self.options.get_safe("with_dcmtk", False)
        cmake.definitions["OSG_WITH_FFMPEG"] = False
        cmake.definitions["OSG_WITH_GSTREAMER"] = False  # self.options.with_gstreamer
        cmake.definitions["OSG_WITH_DIRECTSHOW"] = False
        cmake.definitions["OSG_WITH_SDL"] = False
        cmake.definitions["OSG_WITH_POPPLER"] = False
        cmake.definitions["OSG_WITH_RSVG"] = False
        cmake.definitions["OSG_WITH_NVTT"] = False
        cmake.definitions["OSG_WITH_ASIO"] = False  # self.options.get_safe("with_asio", False)
        cmake.definitions["OSG_WITH_ZEROCONF"] = False
        cmake.definitions["OSG_WITH_LIBLAS"] = False
        cmake.definitions["OSG_WITH_GIF"] = self.options.get_safe("with_gif", False)
        cmake.definitions["OSG_WITH_JPEG"] = self.options.get_safe("with_jpeg", False)
        cmake.definitions["OSG_WITH_PNG"] = self.options.get_safe("with_png", False)
        cmake.definitions["OSG_WITH_TIFF"] = self.options.with_tiff

        if self.settings.os == "Windows":
            # osg has optional quicktime support on Windows
            cmake.definitions["CMAKE_DISABLE_FIND_PACKAGE_QuickTime"] = True

        cmake.definitions["OSG_MSVC_VERSIONED_DLL"] = False

        cmake.configure()
        self._cmake = cmake
        return self._cmake

    def build(self):
        self._patch_sources()
        cmake = self._configure_cmake()
        cmake.build()

    def package(self):
        cmake = self._configure_cmake()
        cmake.install()
```

Reply:
patch in `build()` and cache cmake

Reply:
I generally don't like configuring cmake in the package step. It can cause files to be touched unnecessarily triggering unnecessary rebuilds.

Reply:
unless calling conan create -kb, configure is not triggered again with this modification.

Reply:
Is that guaranteed by conan? I don't recall seeing that in the documentation.

Also, `create -kb` is actually a very useful command for debugging osg's complicated `package_info`, so it would be unfortunate to blow that up.

I guess from my perspective, in the grand scheme of things, caching the cmake build helper is minuscule in both memory and time compared to the process of building osg, so it feels like a needless optimization.

Reply:
There's also the matter of consistency and maintainability which are massively important as C3i continues to grow

Reply:
@prince-chrismc As I mentioned in my previous comment, configuring under the package step is a major blow to the maintainability of this particular package, and I don't think that consistency should be enforced to the detriment of functionality.

Reply:
I've tested to build openscenegraph on Linux, and reconfigure step is very fast (maybe 3-4s when calling `conan create -kb`). So I don't see any reason to not follow the pattern used in all others recipes in CCI. Moreover it won't break if we change `cmake.configure()` arguments.

Reply:
Okay, I gave this a try on Windows, my primary development platform. My recollection was clearly from earlier on in the development when I had been doing more refactoring, yes it seems that `-kb` works reasonably. I'll make the change.